runtime/src/mm/cpp/ExtraObjectData.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "ExtraObjectData.hpp"
+
+#include "PointerBits.h"
+#include "Weak.h"
+
+#ifdef KONAN_OBJC_INTEROP
+#include "ObjCMMAPI.h"
+#endif
+
+using namespace kotlin;
+
+// static
+mm::ExtraObjectData& mm::ExtraObjectData::Install(ObjHeader* object) noexcept {
+    if (object->has_meta_object()) {
+        return mm::ExtraObjectData::FromMetaObjHeader(object->meta_object());
+    }
+
+    TypeInfo* typeInfo = object->typeInfoOrMeta_;
```

Comment:
What if this pointer is already an ExtraObjectData?

Reply:
Right, a race, fixed.

Reply:
Please try to add a test that would detect this race.

Reply:
`ExtraObjectDataTest.ConcurrentInstall` should trigger it. However, according to the assembly (on x64), `TypeInfo* typeInfo = object->typeInfoOrMeta_;` didn't generate another load from `object`, it reused the load from `object->has_meta_object()`, so in practice this race cannot occur (on x64).

Reply:
> it reused the load from `object->has_meta_object()`

Did the latter get inlined?

Reply:
Yes.